README.md
# au.org.greens.trackpaypal

![Screenshot](/images/screenshot.png)

This extension is designed to facilitate tracking of PayPal transactions in Google Analytics

The extension is licensed under [AGPL-3.0](LICENSE.txt).

## Requirements

* PHP v5.4+
* CiviCRM 5.8+
* PayPal account with Instant Payment Notification (IPN) enabled and configured to send data to CiviCRM
* Google Analytics account containing a Property with a tracking code

## Installation (Web UI)

This extension has not yet been published for installation via the web UI.

## Installation (CLI, Zip)

Sysadmins and developers may download the `.zip` file for this extension and
install it with the command-line tool [cv](https://github.com/civicrm/cv).

```bash
cd <extension-dir>
cv dl au.org.greens.trackpaypal@https://github.com/australiangreens/au.org.greens.trackpaypal/archive/master.zip
```

## Installation (CLI, Git)

Sysadmins and developers may clone the [Git](https://en.wikipedia.org/wiki/Git) repo for this extension and
install it with the command-line tool [cv](https://github.com/civicrm/cv).

```bash
git clone https://github.com/australiangreens/au.org.greens.trackpaypal.git
cv en trackpaypal
```

## Usage

Enable the extension then go to the settings page (path is `/civicrm/trackpaypal/settings`; `PayPal Tracking` under the `Administer` menu). Set your preferred options. Be sure you enter your Google Analytics tracking code correctly.

You may wish to start with the extension in Debug Mode. In this mode transaction data will be sent to Google Analytics for recording _and validation_, with the validation results being logged in Civi's [debug logs](https://docs.civicrm.org/dev/en/latest/tools/debugging/#viewing-log-files). You can use this to check that the tracking data is properly constructed as it's being sent to your Google Analytics account.

You may find the [CiviCRM Log Viewer extension](https://civicrm.org/extensions/civicrm-log-viewer) useful for viewing the debug logs.

## Details

This extension rewrites the data packet sent by CiviCRM to PayPal for payment handling to include the Google Analytics ID of the person making the transaction. When PayPal sends its IPN data back to CiviCRM, a hook function parses the returned data and constructs either a Google Analytics [Event](https://developers.google.com/analytics/devguides/collection/analyticsjs/events), [ecommerce transaction](https://developers.google.com/analytics/devguides/collection/analyticsjs/ecommerce) or both. It sends the data using Google's [Measurement Protocol API](https://developers.google.com/analytics/devguides/collection/protocol/v1/).

This extension is not designed to deliver complete Google Analytics tracking coverage of CiviCRM transactions. It was built to specifically track PayPal transactions as they require approaches different to conventional GA tracking.

## Known Issues

N/A


CHANGELOG.md
# Changelog

All notable changes to this extension will be documented in this file.

The extension is licensed under [AGPL-3.0](LICENSE.txt).

## 1.1.4 - 2019-02-05

### Changed

* Updated README.md and screenshot
* Tweaked help text on settings form

### Fixed

* Corrected CiviCRM version compatibility statements in README.md and info.xml

## 1.1.3 - 2019-01-17

### Fixed

* Fixed syntax error

## 1.1.2 - 2019-01-16

### Changed

* Construct fake gcid if none is present in IPN payload to ensure event packets are accepted by GA

## 1.1.1 - 2019-01-15

### Fixed

* Properly log validation response data using `getContents()` method
* Use correct endpoint variable in debug call

## 1.1.0 - 2019-01-15

### Changed

* Changed debug mode behaviour - extension will send data to GA _AND_ the GA validation service

## 1.0.0 - 2019-01-10

### Added

* Can now send ecommerce transaction _and_ event data, or just one or the other
* Can now set a custom Event Category value, defaults to "Transaction"
* Debug mode - send data to Google Analytics hit validation service. Logs return data to Civi's ConfigAndLog

### Changed

* Simpler language for settings page field labels
* Standard event label and value now use Civi form ID and whole dollar amount of the transaction respectively

### Fixed

* Typo in variable name
* Syntax error in category conditional check

## 0.1.0 - 2018-11-06

### Added

* CHANGELOG.md
* Push ecommerce event or standard event to Google Analytics

### Changed

* Rendering of admin form (from @seamuslee001)

## 0.0.6 - 2018-11-05

### Added

* Admin form for extension settings (incomplete)
* Menu item for admin form


trackpaypal.civix.php
<?php

// AUTO-GENERATED FILE -- Civix may overwrite any changes made to this file

/**
 * The ExtensionUtil class provides small stubs for accessing resources of this
 * extension.
 */
class CRM_Trackpaypal_ExtensionUtil {
  const SHORT_NAME = "trackpaypal";
  const LONG_NAME = "au.org.greens.trackpaypal";
  const CLASS_PREFIX = "CRM_Trackpaypal";

  /**
   * Translate a string using the extension's domain.
   *
   * If the extension doesn't have a specific translation
   * for the string, fallback to the default translations.
   *
   * @param string $text
   *   Canonical message text (generally en_US).
   * @param array $params
   * @return string
   *   Translated text.
   * @see ts
   */
  public static function ts($text, $params = array()) {
    if (!array_key_exists('domain', $params)) {
      $params['domain'] = array(self::LONG_NAME, NULL);
    }
    return ts($text, $params);
  }

  /**
   * Get the URL of a resource file (in this extension).
   *
   * @param string|NULL $file
   *   Ex: NULL.
   *   Ex: 'css/foo.css'.
   * @return string
   *   Ex: 'http://example.org/sites/default/ext/org.example.foo'.
   *   Ex: 'http://example.org/sites/default/ext/org.example.foo/css/foo.css'.
   */
  public static function url($file = NULL) {
    if ($file === NULL) {
      return rtrim(CRM_Core_Resources::singleton()->getUrl(self::LONG_NAME), '/');
    }
    return CRM_Core_Resources::singleton()->getUrl(self::LONG_NAME, $file);
  }

  /**
   * Get the path of a resource file (in this extension).
   *
   * @param string|NULL $file
   *   Ex: NULL.
   *   Ex: 'css/foo.css'.
   * @return string
   *   Ex: '/var/www/example.org/sites/default/ext/org.example.foo'.
   *   Ex: '/var/www/example.org/sites/default/ext/org.example.foo/css/foo.css'.
   */
  public static function path($file = NULL) {
    // return CRM_Core_Resources::singleton()->getPath(self::LONG_NAME, $file);
    return __DIR__ . ($file === NULL ? '' : (DIRECTORY_SEPARATOR . $file));
  }

  /**
   * Get the name of a class within this extension.
   *
   * @param string $suffix
   *   Ex: 'Page_HelloWorld' or 'Page\\HelloWorld'.
   * @return string
   *   Ex: 'CRM_Foo_Page_HelloWorld'.
   */
  public static function findClass($suffix) {
    return self::CLASS_PREFIX . '_' . str_replace('\\', '_', $suffix);
  }

}

use CRM_Trackpaypal_ExtensionUtil as E;

/**
 * (Delegated) Implements hook_civicrm_config().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_config
 */
function _trackpaypal_civix_civicrm_config(&$config = NULL) {
  static $configured = FALSE;
  if ($configured) {
    return;
  }
  $configured = TRUE;

  $template =& CRM_Core_Smarty::singleton();

  $extRoot = dirname(__FILE__) . DIRECTORY_SEPARATOR;
  $extDir = $extRoot . 'templates';

  if (is_array($template->template_dir)) {
    array_unshift($template->template_dir, $extDir);
  }
  else {
    $template->template_dir = array($extDir, $template->template_dir);
  }

  $include_path = $extRoot . PATH_SEPARATOR . get_include_path();
  set_include_path($include_path);
}

/**
 * (Delegated) Implements hook_civicrm_xmlMenu().
 *
 * @param $files array(string)
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_xmlMenu
 */
function _trackpaypal_civix_civicrm_xmlMenu(&$files) {
  foreach (_trackpaypal_civix_glob(__DIR__ . '/xml/Menu/*.xml') as $file) {
    $files[] = $file;
  }
}

/**
 * Implements hook_civicrm_install().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_install
 */
function _trackpaypal_civix_civicrm_install() {
  _trackpaypal_civix_civicrm_config();
  if ($upgrader = _trackpaypal_civix_upgrader()) {
    $upgrader->onInstall();
  }
}

/**
 * Implements hook_civicrm_postInstall().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_postInstall
 */
function _trackpaypal_civix_civicrm_postInstall() {
  _trackpaypal_civix_civicrm_config();
  if ($upgrader = _trackpaypal_civix_upgrader()) {
    if (is_callable(array($upgrader, 'onPostInstall'))) {
      $upgrader->onPostInstall();
    }
  }
}

/**
 * Implements hook_civicrm_uninstall().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_uninstall
 */
function _trackpaypal_civix_civicrm_uninstall() {
  _trackpaypal_civix_civicrm_config();
  if ($upgrader = _trackpaypal_civix_upgrader()) {
    $upgrader->onUninstall();
  }
}

/**
 * (Delegated) Implements hook_civicrm_enable().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_enable
 */
function _trackpaypal_civix_civicrm_enable() {
  _trackpaypal_civix_civicrm_config();
  if ($upgrader = _trackpaypal_civix_upgrader()) {
    if (is_callable(array($upgrader, 'onEnable'))) {
      $upgrader->onEnable();
    }
  }
}

/**
 * (Delegated) Implements hook_civicrm_disable().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_disable
 * @return mixed
 */
function _trackpaypal_civix_civicrm_disable() {
  _trackpaypal_civix_civicrm_config();
  if ($upgrader = _trackpaypal_civix_upgrader()) {
    if (is_callable(array($upgrader, 'onDisable'))) {
      $upgrader->onDisable();
    }
  }
}

/**
 * (Delegated) Implements hook_civicrm_upgrade().
 *
 * @param $op string, the type of operation being performed; 'check' or 'enqueue'
 * @param $queue CRM_Queue_Queue, (for 'enqueue') the modifiable list of pending up upgrade tasks
 *
 * @return mixed  based on op. for 'check', returns array(boolean) (TRUE if upgrades are pending)
 *                for 'enqueue', returns void
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_upgrade
 */
function _trackpaypal_civix_civicrm_upgrade($op, CRM_Queue_Queue $queue = NULL) {
  if ($upgrader = _trackpaypal_civix_upgrader()) {
    return $upgrader->onUpgrade($op, $queue);
  }
}

/**
 * @return CRM_Trackpaypal_Upgrader
 */
function _trackpaypal_civix_upgrader() {
  if (!file_exists(__DIR__ . '/CRM/Trackpaypal/Upgrader.php')) {
    return NULL;
  }
  else {
    return CRM_Trackpaypal_Upgrader_Base::instance();
  }
}

/**
 * Search directory tree for files which match a glob pattern
 *
 * Note: Dot-directories (like "..", ".git", or ".svn") will be ignored.
 * Note: In Civi 4.3+, delegate to CRM_Utils_File::findFiles()
 *
 * @param $dir string, base dir
 * @param $pattern string, glob pattern, eg "*.txt"
 * @return array(string)
 */
function _trackpaypal_civix_find_files($dir, $pattern) {
  if (is_callable(array('CRM_Utils_File', 'findFiles'))) {
    return CRM_Utils_File::findFiles($dir, $pattern);
  }

  $todos = array($dir);
  $result = array();
  while (!empty($todos)) {
    $subdir = array_shift($todos);
    foreach (_trackpaypal_civix_glob("$subdir/$pattern") as $match) {
      if (!is_dir($match)) {
        $result[] = $match;
      }
    }
    if ($dh = opendir($subdir)) {
      while (FALSE !== ($entry = readdir($dh))) {
        $path = $subdir . DIRECTORY_SEPARATOR . $entry;
        if ($entry{0} == '.') {
        }
        elseif (is_dir($path)) {
          $todos[] = $path;
        }
      }
      closedir($dh);
    }
  }
  return $result;
}
/**
 * (Delegated) Implements hook_civicrm_managed().
 *
 * Find any *.mgd.php files, merge their content, and return.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_managed
 */
function _trackpaypal_civix_civicrm_managed(&$entities) {
  $mgdFiles = _trackpaypal_civix_find_files(__DIR__, '*.mgd.php');
  foreach ($mgdFiles as $file) {
    $es = include $file;
    foreach ($es as $e) {
      if (empty($e['module'])) {
        $e['module'] = E::LONG_NAME;
      }
      $entities[] = $e;
      if (empty($e['params']['version'])) {
        $e['params']['version'] = '3';
      }
    }
  }
}

/**
 * (Delegated) Implements hook_civicrm_caseTypes().
 *
 * Find any and return any files matching "xml/case/*.xml"
 *
 * Note: This hook only runs in CiviCRM 4.4+.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_caseTypes
 */
function _trackpaypal_civix_civicrm_caseTypes(&$caseTypes) {
  if (!is_dir(__DIR__ . '/xml/case')) {
    return;
  }

  foreach (_trackpaypal_civix_glob(__DIR__ . '/xml/case/*.xml') as $file) {
    $name = preg_replace('/\.xml$/', '', basename($file));
    if ($name != CRM_Case_XMLProcessor::mungeCaseType($name)) {
      $errorMessage = sprintf("Case-type file name is malformed (%s vs %s)", $name, CRM_Case_XMLProcessor::mungeCaseType($name));
      CRM_Core_Error::fatal($errorMessage);
      // throw new CRM_Core_Exception($errorMessage);
    }
    $caseTypes[$name] = array(
      'module' => E::LONG_NAME,
      'name' => $name,
      'file' => $file,
    );
  }
}

/**
 * (Delegated) Implements hook_civicrm_angularModules().
 *
 * Find any and return any files matching "ang/*.ang.php"
 *
 * Note: This hook only runs in CiviCRM 4.5+.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_angularModules
 */
function _trackpaypal_civix_civicrm_angularModules(&$angularModules) {
  if (!is_dir(__DIR__ . '/ang')) {
    return;
  }

  $files = _trackpaypal_civix_glob(__DIR__ . '/ang/*.ang.php');
  foreach ($files as $file) {
    $name = preg_replace(':\.ang\.php$:', '', basename($file));
    $module = include $file;
    if (empty($module['ext'])) {
      $module['ext'] = E::LONG_NAME;
    }
    $angularModules[$name] = $module;
  }
}

/**
 * Glob wrapper which is guaranteed to return an array.
 *
 * The documentation for glob() says, "On some systems it is impossible to
 * distinguish between empty match and an error." Anecdotally, the return
 * result for an empty match is sometimes array() and sometimes FALSE.
 * This wrapper provides consistency.
 *
 * @link http://php.net/glob
 * @param string $pattern
 * @return array, possibly empty
 */
function _trackpaypal_civix_glob($pattern) {
  $result = glob($pattern);
  return is_array($result) ? $result : array();
}

/**
 * Inserts a navigation menu item at a given place in the hierarchy.
 *
 * @param array $menu - menu hierarchy
 * @param string $path - path to parent of this item, e.g. 'my_extension/submenu'
 *    'Mailing', or 'Administer/System Settings'
 * @param array $item - the item to insert (parent/child attributes will be
 *    filled for you)
 */
function _trackpaypal_civix_insert_navigation_menu(&$menu, $path, $item) {
  // If we are done going down the path, insert menu
  if (empty($path)) {
    $menu[] = array(
      'attributes' => array_merge(array(
        'label'      => CRM_Utils_Array::value('name', $item),
        'active'     => 1,
      ), $item),
    );
    return TRUE;
  }
  else {
    // Find an recurse into the next level down
    $found = FALSE;
    $path = explode('/', $path);
    $first = array_shift($path);
    foreach ($menu as $key => &$entry) {
      if ($entry['attributes']['name'] == $first) {
        if (!isset($entry['child'])) {
          $entry['child'] = array();
        }
        $found = _trackpaypal_civix_insert_navigation_menu($entry['child'], implode('/', $path), $item, $key);
      }
    }
    return $found;
  }
}

/**
 * (Delegated) Implements hook_civicrm_navigationMenu().
 */
function _trackpaypal_civix_navigationMenu(&$nodes) {
  if (!is_callable(array('CRM_Core_BAO_Navigation', 'fixNavigationMenu'))) {
    _trackpaypal_civix_fixNavigationMenu($nodes);
  }
}

/**
 * Given a navigation menu, generate navIDs for any items which are
 * missing them.
 */
function _trackpaypal_civix_fixNavigationMenu(&$nodes) {
  $maxNavID = 1;
  array_walk_recursive($nodes, function($item, $key) use (&$maxNavID) {
    if ($key === 'navID') {
      $maxNavID = max($maxNavID, $item);
    }
  });
  _trackpaypal_civix_fixNavigationMenuItems($nodes, $maxNavID, NULL);
}

function _trackpaypal_civix_fixNavigationMenuItems(&$nodes, &$maxNavID, $parentID) {
  $origKeys = array_keys($nodes);
  foreach ($origKeys as $origKey) {
    if (!isset($nodes[$origKey]['attributes']['parentID']) && $parentID !== NULL) {
      $nodes[$origKey]['attributes']['parentID'] = $parentID;
    }
    // If no navID, then assign navID and fix key.
    if (!isset($nodes[$origKey]['attributes']['navID'])) {
      $newKey = ++$maxNavID;
      $nodes[$origKey]['attributes']['navID'] = $newKey;
      $nodes[$newKey] = $nodes[$origKey];
      unset($nodes[$origKey]);
      $origKey = $newKey;
    }
    if (isset($nodes[$origKey]['child']) && is_array($nodes[$origKey]['child'])) {
      _trackpaypal_civix_fixNavigationMenuItems($nodes[$origKey]['child'], $maxNavID, $nodes[$origKey]['attributes']['navID']);
    }
  }
}

/**
 * (Delegated) Implements hook_civicrm_alterSettingsFolders().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_alterSettingsFolders
 */
function _trackpaypal_civix_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
  static $configured = FALSE;
  if ($configured) {
    return;
  }
  $configured = TRUE;

  $settingsDir = __DIR__ . DIRECTORY_SEPARATOR . 'settings';
  if (is_dir($settingsDir) && !in_array($settingsDir, $metaDataFolders)) {
    $metaDataFolders[] = $settingsDir;
  }
}

/**
 * (Delegated) Implements hook_civicrm_entityTypes().
 *
 * Find any *.entityType.php files, merge their content, and return.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_entityTypes
 */

function _trackpaypal_civix_civicrm_entityTypes(&$entityTypes) {
  $entityTypes = array_merge($entityTypes, array (
  ));
}


settings/Trackpaypal.setting.php
<?php
/*
 +--------------------------------------------------------------------+
 | CiviCRM version 4.7                                                |
 +--------------------------------------------------------------------+
 | Copyright CiviCRM LLC (c) 2004-2017                                |
 +--------------------------------------------------------------------+
 | This file is a part of CiviCRM.                                    |
 |                                                                    |
 | CiviCRM is free software; you can copy, modify, and distribute it  |
 | under the terms of the GNU Affero General Public License           |
 | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
 |                                                                    |
 | CiviCRM is distributed in the hope that it will be useful, but     |
 | WITHOUT ANY WARRANTY; without even the implied warranty of         |
 | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
 | See the GNU Affero General Public License for more details.        |
 |                                                                    |
 | You should have received a copy of the GNU Affero General Public   |
 | License and the CiviCRM Licensing Exception along                  |
 | with this program; if not, contact CiviCRM LLC                     |
 | at info[AT]civicrm[DOT]org. If you have questions about the        |
 | GNU Affero General Public License or the licensing of CiviCRM,     |
 | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
 +--------------------------------------------------------------------+
 */
/**
 *
 * @package CRM
 * @copyright CiviCRM LLC (c) 2004-2017
 * $Id$
 *
 */
/*
 * Settings metadata file
 */
return array(
  'trackpaypal_event_type' => array(
    'group_name' => 'trackpaypal',
    'group' => 'trackpaypal',
    'name' => 'trackpaypal_event_type',
    'filter' => 'trackpaypal',
    'type' => 'String',
    'add' => '4.7',
    'is_domain' => 1,
    'is_contact' => 0,
    'description' => 'Send eCommerce Transaction, standard Event or both to Google Analytics?',
    'title' => 'Data to send to GA',
    'default' => array('ecommerce'),
    'html_type' => 'Select',
    'pseudoconstant' => array(
      'callback' => 'CRM_Trackpaypal_Form_Settings::eventTypes',
    ),
    'quick_form_type' => 'Element',
  ),
  'trackpaypal_tracking_code' => array(
    'group_name' => 'trackpaypal',
    'group' => 'trackpaypal',
    'name' => 'trackpaypal_tracking_code',
    'filter' => 'trackpaypal',
    'type' => 'String',
    'add' => '4.7',
    'is_domain' => 1,
    'is_contact' => 0,
    'description' => 'Google Analytics tracking code (similar to UA-XXXXXX-X)',
    'title' => 'Tracking Code',
    'html_type' => 'Text',
    'html_attributes' => array(
      'size' => 18,
      'maxlength' => 16,
    ),
    'quick_form_type' => 'Element',
  ),
  'trackpaypal_event_category' => array(
    'group_name' => 'trackpaypal',
    'group' => 'trackpaypal',
    'name' => 'trackpaypal_event_category',
    'filter' => 'trackpaypal',
    'type' => 'String',
    'add' => '4.7',
    'is_domain' => 1,
    'is_contact' => 0,
    'description' => 'Google Analytics Event Category term. Defaults to "Transaction" if blank.',
    'title' => 'Event Category',
    'html_type' => 'Text',
    'html_attributes' => array(
      'size' => 18,
      'maxlength' => 16,
    ),
    'quick_form_type' => 'Element',
  ),
  'trackpaypal_debug_mode' => array(
    'group_name' => 'trackpaypal',
    'group' => 'trackpaypal',
    'name' => 'trackpaypal_debug_mode',
    'filter' => 'trackpaypal',
    'type' => 'String',
    'add' => '4.7',
    'is_domain' => 1,
    'is_contact' => 0,
    'description' => 'Enable debug mode? Logs GA validation data to ConfigAndLog. Data will sent to GA as well.',
    'title' => 'Debug mode',
    'default' => array('off'),
    'html_type' => 'Select',
    'pseudoconstant' => array(
      'callback' => 'CRM_Trackpaypal_Form_Settings::debugMode',
    ),
    'quick_form_type' => 'Element',
  ),

);


trackpaypal.php
<?php

require_once 'trackpaypal.civix.php';
use CRM_Trackpaypal_ExtensionUtil as E;

/**
 * Implements hook_civicrm_config().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_config
 */
function trackpaypal_civicrm_config(&$config) {
  _trackpaypal_civix_civicrm_config($config);
}

/**
 * Implements hook_civicrm_xmlMenu().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_xmlMenu
 */
function trackpaypal_civicrm_xmlMenu(&$files) {
  _trackpaypal_civix_civicrm_xmlMenu($files);
}

/**
 * Implements hook_civicrm_install().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_install
 */
function trackpaypal_civicrm_install() {
  _trackpaypal_civix_civicrm_install();
}

/**
 * Implements hook_civicrm_postInstall().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_postInstall
 */
function trackpaypal_civicrm_postInstall() {
  _trackpaypal_civix_civicrm_postInstall();
}

/**
 * Implements hook_civicrm_uninstall().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_uninstall
 */
function trackpaypal_civicrm_uninstall() {
  _trackpaypal_civix_civicrm_uninstall();
}

/**
 * Implements hook_civicrm_enable().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_enable
 */
function trackpaypal_civicrm_enable() {
  _trackpaypal_civix_civicrm_enable();
}

/**
 * Implements hook_civicrm_disable().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_disable
 */
function trackpaypal_civicrm_disable() {
  _trackpaypal_civix_civicrm_disable();
}

/**
 * Implements hook_civicrm_upgrade().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_upgrade
 */
function trackpaypal_civicrm_upgrade($op, CRM_Queue_Queue $queue = NULL) {
  return _trackpaypal_civix_civicrm_upgrade($op, $queue);
}

/**
 * Implements hook_civicrm_managed().
 *
 * Generate a list of entities to create/deactivate/delete when this module
 * is installed, disabled, uninstalled.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_managed
 */
function trackpaypal_civicrm_managed(&$entities) {
  _trackpaypal_civix_civicrm_managed($entities);
}

/**
 * Implements hook_civicrm_caseTypes().
 *
 * Generate a list of case-types.
 *
 * Note: This hook only runs in CiviCRM 4.4+.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_caseTypes
 */
function trackpaypal_civicrm_caseTypes(&$caseTypes) {
  _trackpaypal_civix_civicrm_caseTypes($caseTypes);
}

/**
 * Implements hook_civicrm_angularModules().
 *
 * Generate a list of Angular modules.
 *
 * Note: This hook only runs in CiviCRM 4.5+. It may
 * use features only available in v4.6+.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_angularModules
 */
function trackpaypal_civicrm_angularModules(&$angularModules) {
  _trackpaypal_civix_civicrm_angularModules($angularModules);
}

/**
 * Implements hook_civicrm_alterSettingsFolders().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_alterSettingsFolders
 */
function trackpaypal_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
  _trackpaypal_civix_civicrm_alterSettingsFolders($metaDataFolders);
}

/**
 * Implements hook_civicrm_entityTypes().
 *
 * Declare entity types provided by this module.
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_entityTypes
 */
function trackpaypal_civicrm_entityTypes(&$entityTypes) {
  _trackpaypal_civix_civicrm_entityTypes($entityTypes);
}

// --- Functions below this ship commented out. Uncomment as required. ---

/**
 * Implements hook_civicrm_preProcess().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_preProcess
 *
function trackpaypal_civicrm_preProcess($formName, &$form) {

} // */

/**
 * Implements hook_civicrm_navigationMenu().
 *
 * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_navigationMenu
 *
 */
function trackpaypal_civicrm_navigationMenu(&$menu) {
  _trackpaypal_civix_insert_navigation_menu($menu, 'Administer', array(
    'label' => E::ts('PayPal Tracking'),
    'name' => 'trackpaypal_settings',
    'url' => 'civicrm/trackpaypal/settings',
    'permission' => 'administer CiviCRM',
    'operator' => 'OR',
    'separator' => 0,
  ));
  _trackpaypal_civix_navigationMenu($menu);
}

/**
 * Implements hook_civicrm_buildForm().
 *
 * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_buildForm/
 */
function trackpaypal_civicrm_buildForm($formName, &$form) {
  if ($formName === 'CRM_Contribute_Form_Contribution_Main'
    || $formName === 'CRM_Contribute_Form_Contribution_Confirm') {

    // Add a hidden field to the form with the id/key 'gcid'
    // JavaScript code in the template will populate it
    // with the Google Analytics client from the visitor's browser
    $templatePath = realpath(dirname(__FILE__) . "/templates");
    $form->add('hidden', 'gcid', '');
    CRM_Core_Region::instance('page-body')->add(array(
      'template' => "{$templatePath}/trackpaypal.tpl",
    ));
  }
}

/**
 * Implements hook_civicrm_alterPaymentProcessorParams().
 *
 * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_alterPaymentProcessorParams/
 */
function trackpaypal_civicrm_alterPaymentProcessorParams($paymentObj,&$rawParams, &$cookedParams) {
  // Eway Payment Processor sometimes passes $cookedParams as an instance of GatewayRequest class
  // PHP complains if you try and use it as an array and we don't care about it in this instnce so return.
  if ($cookedParams instanceof GatewayRequest) {
    return;
  }
  else {
    if (isset($cookedParams['custom'])) {
      // Add the Google Analytics client ID value
      // to the JSON encoded 'custom' attribute
      // of the payload that goes to PayPal
      $customPayload = json_decode($cookedParams['custom'], TRUE);
      $customPayload += ['gcid' => $rawParams['gcid']];
      $cookedParams['custom'] = json_encode($customPayload);
    }
  }
}

function trackpaypal_civicrm_postIPNProcess(&$IPNData) {
  // We've captured the IPN packet
  // Now we want to retrieve the Google Client ID
  // and transaction details to send to Google Analytics
  // via its REST interface

  // Retrieve extension settings
    $event_type = Civi::settings()->get('trackpaypal_event_type');
    $tracking_code = Civi::settings()->get('trackpaypal_tracking_code');
    $event_category = Civi::settings()->get('trackpaypal_event_category');
    if ($event_category == "") { $event_category = 'Transaction'; }
    $debug_mode = Civi::settings()->get('trackpaypal_debug_mode');

  // Check the GA Code is of valid syntax
  // If not we do nothing
  if (!trackpaypal_isGACode($tracking_code)) {
    Civi::log()->warning('GA Code invalid: {code}', array('code' => $tracking_code));
    return;
  }

  // Retrieve IPN packet data
  // We're constructing a fake $gcid if none is present in the payload
  $customPayload = json_decode($IPNData['custom'], TRUE);
  $gcid = ($customPayload['gcid'] ? $customPayload['gcid'] : mt_rand(100000000,999999999) . '.' . time());
  $form_id = $customPayload['contributionPageID'];
  $trxn_id = $IPNData['txn_id'];
  $revenue = $IPNData['mc_gross'];
  $currency = $IPNData['mc_currency'];

  // Construct HTTP request object
  $client = new GuzzleHttp\Client(['base_uri' => 'https://www.google-analytics.com']);

  // Define our endpoints (production and debug)
  $endpoint = '/collect';
  $endpoint_debug = '/debug/collect';

  // Construct our two payload types: standard GA event and ecommerce transaction

  $packet_ecommerce = [
    'form_params' => [
        'v' => '1',
        'tid' => $tracking_code,
        'cid' => $gcid,
        't' => 'transaction',
        'ti' => $trxn_id,
        'tr' => $revenue,
        'cu' => $currency,
      ]
    ];

  $packet_event = [
    'form_params' => [
        'v' => '1',
        'tid' => $tracking_code,
        'cid' => $gcid,
        't' => 'event',
        'ec' => $event_category,
        'ea' => 'purchase',
        'el' => $form_id,
        'ev' => floor(floatval($revenue)),
      ]
    ];

  if ($event_type == 'ecommerce') {
    $result = $client->request('POST', $endpoint, $packet_ecommerce);
    if ($debug_mode == 'on') {
      $result = $client->request('POST', $endpoint_debug, $packet_ecommerce);
      trackpaypal_logValidation($result);
    }
  }
  else if ($event_type == 'standard') {
    $result = $client->request('POST', $endpoint, $packet_event);
    if ($debug_mode == 'on') {
      $result = $client->request('POST', $endpoint_debug, $packet_event);
      trackpaypal_logValidation($result);
    }
  }
  else if ($event_type == 'both') {
    $result = $client->request('POST', $endpoint, $packet_ecommerce);
    if ($debug_mode == 'on') {
      $result = $client->request('POST', $endpoint_debug, $packet_ecommerce);
      trackpaypal_logValidation($result);
    }
    $result = $client->request('POST', $endpoint, $packet_event);
    if ($debug_mode == 'on') {
      $result = $client->request('POST', $endpoint_debug, $packet_event);
      trackpaypal_logValidation($result);
    }
  }
}

function trackpaypal_isGACode($str) {
  return (bool) preg_match('/^ua-\d{4,10}(-\d{1,4})?$/i', $str);
}

function trackpaypal_logValidation($response) {
  if ($response->getBody()) {
    Civi::log()->debug('GA validation response: {response}', array(
      'response' => $response->getBody()->getContents(),
    ));
  }
}

